### PR TITLE
ENH: add `fetch-eggnog-hmmer-db` action

### LINKED ISSUE #167: ENH: add `fetch-hmmer-db` action
Allow orthologue prediction with a HMMER database through eggNOG. 

```[tasklist]
#### Depends on
- [ ] qiime2/q2-types#327
```

### PR DESCRIPTION
**Pull request by Sann5**
### What's new
- An action that fetches a HMMER database. This will be used downstream to find seed orthologues (to be implemented).
- Closes #167
- ❌ CI tests should fail until a version of q2-types with the PR mentioned below is up in conda

```[tasklist]
### Blocked by...
- [ ] depends on qiime2/q2-types#328
```

### Run it locally 
1. Checkout the PR branch.

> Assuming 1) you already have a local copy of `q2-moshpit` that is installed in editable mode in your activated virtual environment and 2) that "origin" is the name of the remote of this GitHub repo; run the following. 

```bash
cd my_q2-moshpit
PR=173
git fetch origin pull/${PR}/head:pr-${PR}
git checkout pr-${PR}
```
> If the PR in q2-types mentioned above is not yet merged you will need to do the same for q2-types. E.g:

```bash
cd my_q2-types
PR=328
git fetch origin pull/${PR}/head:pr-${PR}
git checkout pr-${PR}
```

2. Test it out!
```bash
qiime moshpit fetch-eggnog-hmmer-db --p-taxon-id 1100069 --output-dir 1100069_db --verbose
```

### CODE REVIEW

**Comment by VinzentRisch** [comment]
Hi @Sann5 
I don't know if i have time to look at it tomorrow. Maybe I'll only look at it next week. Hope that is not a problem :)

File: q2_moshpit/eggnog/_utils.py
Comment (VinzentRisch) [inline q2_moshpit/eggnog/_utils.py:77]: Sorry this is a bit nit-picky, but these quotes should all be consistent with the rest of the file.

```suggestion
                                        r"\.faa\.final_tree(\.fa)?", "", line
```

File: q2_moshpit/eggnog/_utils.py
Comment (VinzentRisch) [inline q2_moshpit/eggnog/_utils.py:117]:
```suggestion
            if f.endswith(".gz")
```

File: q2_moshpit/eggnog/_utils.py
Comment (VinzentRisch) [inline q2_moshpit/eggnog/_utils.py:125]:
```suggestion
            with gzip.open(fpi, "rt") as f_in, open(fpo, "w") as f_out:
```

File: q2_moshpit/eggnog/_utils.py
Comment (VinzentRisch) [inline q2_moshpit/eggnog/_utils.py:127]:
```suggestion
                content = content.replace("-", "")
```

**Review by VinzentRisch — CHANGES_REQUESTED** [review]
Looks very good!
I ran it and it works with no problems.
See some minor changes and typos below. :)

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:77]: no worries :)

**Review by VinzentRisch — APPROVED** [review]
Everything looks good! :)

**Comment by Sann5** [comment]
I made quite a bit of changes @VinzentRisch since there were changes required in the dependent PR. could you give it a another look before i merge?

File: q2_moshpit/eggnog/_utils.py
Comment (misialq) [inline q2_moshpit/eggnog/_utils.py:81]: This will read the entire file into memory - there's no need for that. You should just read line-by-line and either modify and write into the new file or just write.

File: q2_moshpit/eggnog/_utils.py
Comment (misialq) [inline q2_moshpit/eggnog/_utils.py:70]: Please extract this entire context into a testable function and write some unit tests - it will make the parent function much more readable.

File: q2_moshpit/plugin_setup.py
Comment (misialq) [inline q2_moshpit/plugin_setup.py:1489]: Let's call this output `seed-alignments`.

File: q2_moshpit/plugin_setup.py
Comment (misialq) [inline q2_moshpit/plugin_setup.py:1473]: When I try to run it with `1` I get an error:
```
Validating taxon ID:
Downloading taxonomy file...
Running external command line application(s). This may print messages to stdout and/or stderr.
The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.

Command: wget -O /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpkuvn0wmy/e5.taxid_info.tsv http://eggnog5.embl.de/download/eggnog_5.0/e5.taxid_info.tsv

--2024-06-20 13:08:45--  http://eggnog5.embl.de/download/eggnog_5.0/e5.taxid_info.tsv
Resolving eggnog5.embl.de (eggnog5.embl.de)... 194.94.44.170
Connecting to eggnog5.embl.de (eggnog5.embl.de)|194.94.44.170|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1429293 (1.4M) [application/octet-stream]
Saving to: ‘/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpkuvn0wmy/e5.taxid_info.tsv’

/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpkuvn0wmy/e5.taxid_info.tsv               100%[============================================================================================================================================================================================================================================>]   1.36M  1.17MB/s    in 1.2s

2024-06-20 13:08:47 (1.17 MB/s) - ‘/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpkuvn0wmy/e5.taxid_info.tsv’ saved [1429293/1429293]

Valid taxon ID.
Proceeding with HMMER database download and build...
Running external command line application(s). This may print messages to stdout and/or stderr.
The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.

Command: wget -O /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpxaz500ny/1_hmms.tar.gz http://eggnog5.embl.de/download/eggnog_5.0/per_tax_level/1/1_hmms.tar.gz

--2024-06-20 13:08:47--  http://eggnog5.embl.de/download/eggnog_5.0/per_tax_level/1/1_hmms.tar.gz
Resolving eggnog5.embl.de (eggnog5.embl.de)... 194.94.44.170
Connecting to eggnog5.embl.de (eggnog5.embl.de)|194.94.44.170|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 56 [application/octet-stream]
Saving to: ‘/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpxaz500ny/1_hmms.tar.gz’

/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpxaz500ny/1_hmms.tar.gz                   100%[============================================================================================================================================================================================================================================>]      56  --.-KB/s    in 0s

2024-06-20 13:08:47 (4.86 MB/s) - ‘/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/tmpxaz500ny/1_hmms.tar.gz’ saved [56/56]

Decompressing...
Running external command line application(s). This may print messages to stdout and/or stderr.
The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.

Command: tar zxf 1_hmms.tar.gz

Merging hmm files...
Preparing HMM database...
Running external command line application(s). This may print messages to stdout and/or stderr.
The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.

Command: hmmpress /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-PressedProfileHmmsDirectoryFmt-0kxly8rr/1.hmm


Error: File format problem in trying to open HMM file /var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-PressedProfileHmmsDirectoryFmt-0kxly8rr/1.hmm.
File exists, but appears to be empty?

Traceback (most recent call last):
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/q2cli/commands.py", line 520, in __call__
    results = self._execute_action(
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/q2cli/commands.py", line 581, in _execute_action
    results = action(**arguments)
  File "<decorator-gen-758>", line 2, in fetch_eggnog_hmmer_db
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/qiime2/sdk/action.py", line 342, in bound_callable
    outputs = self._callable_executor_(
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/qiime2/sdk/action.py", line 576, in _callable_executor_
    output_views = self._callable(**view_args)
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/q2_moshpit/eggnog/_dbs.py", line 362, in fetch_eggnog_hmmer_db
    idmap, hmmer_db, pressed_hmmer_db = _download_and_build_hmm_db(taxon_id)
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/q2_moshpit/eggnog/_utils.py", line 103, in _download_and_build_hmm_db
    run_command(cmd=["hmmpress", hmms_merged_p])
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/site-packages/q2_moshpit/_utils.py", line 37, in run_command
    subprocess.run(cmd, check=True, **kwargs)
  File "/Users/mziemski/miniconda3/envs/q2-metagenome-2024.5/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['hmmpress', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-PressedProfileHmmsDirectoryFmt-0kxly8rr/1.hmm']' returned non-zero exit status 1.

Plugin error from moshpit:

  Command '['hmmpress', '/var/folders/7f/7nw_x13n5q965rss_qz6061m0000gq/T/q2-PressedProfileHmmsDirectoryFmt-0kxly8rr/1.hmm']' returned non-zero exit status 1.

See above for debug info.
```
Does not happen with `2` (or any other I tried).

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @Sann5, I think this still needs a bit of work - see below.

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:150]: I think this function might save quite a few lines of code if we want to use it in the rest of the wget commands in the _dbs.py module. I think this would be good practice since until now we are not really supporting the case in wich the download fails. But perhaps its better to leave that to for MAINT PR.

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:55]: What I did in order to prevent this is that the first thing that happens is that the taxon id is validated. So this should never happen as long as `_download_and_build_hmm_db` is never called directly. If you try to run the action specifying 8 as taxon id you would get 

```
'8' is not valid taxon ID. To view all valid taxon IDs inspect e5.taxid_info.tsv. You can download it with this command: wget http://eggnog5.embl.de/download/eggnog_5.0/e5.taxid_info.tsv
```

Comment (misialq) [reply q2_moshpit/eggnog/_utils.py:55]: Well, clearly it does happen. First of all, what your function checks is whether the taxon ID is present in the taxid_info file (which it should!) - what I am referring to is those being present/absent on the server. It's great it works with taxon 8 but that's not the taxon I used - mine was `1773` (in which case, your message is even more confusing: what does the 8 even mean there?). Please build in exception handling at the download level to avoid this kind of situation.

File: q2_moshpit/eggnog/_utils.py
Comment (misialq) [reply q2_moshpit/eggnog/_utils.py:150]: If you see my comment below on handling the failed download, it would give even more support to just placing this code directly where this function is being called. Yes, it will save you 3-4 lines of code, probably not such a big deal when you contrast it with the visibility and a more custom exception-handling. All the URLs you're using have a very long common prefix - you can just extract it into a constant and place it on top of the module - will save you some lines too ;)

File: q2_moshpit/plugin_setup.py
Comment (Sann5) [reply q2_moshpit/plugin_setup.py:1473]: @misialq waa... this is weird. 1 is the root of the taxonomic tree no so it should essentially contain every taxa. Instead `1_hmms.tar.gz` is empty. But it exists in the ftp server. I would just say we limit the range from 2 onwards.

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:55]: Oh, I see what you mean now. My mistake. I thought if it was on the taxid_info file it's on the FTP server, but that is not true. 8 is the return code from the failed wget command, which apparently means "Server issued an error response", which is funny enough not so much more informative than the code itself 😅. I will...

1. Change the error message so it says, "error code: x"
2. change the validation so it actually checks whether the taxon id is on the server rather than on the taxid_info file.

Comment (misialq) [reply q2_moshpit/eggnog/_utils.py:55]: Hmm, I wonder for point 2: maybe you should actually do both, just to be sure that the id is present everywhere where it should be...

Comment (misialq) [reply q2_moshpit/eggnog/_utils.py]: Yeah, that sounds right 🙌

File: q2_moshpit/eggnog/_utils.py
Comment (misialq) [inline q2_moshpit/eggnog/_utils.py:153]:
```suggestion
                    if re.search(r'(\d+)/', attr[1]):
                        self.links.append(match.group(1))
```

File: q2_moshpit/plugin_setup.py
Comment (misialq) [inline q2_moshpit/plugin_setup.py:1482]: Hmmm, looks like we forgot to test the output name change? :P
```suggestion
        ("seed_alignments", GenomeData[Proteins] % Properties("eggnog"))
```

File: q2_moshpit/plugin_setup.py
Comment (misialq) [inline q2_moshpit/plugin_setup.py:1489]:
```suggestion
        "seed_alignments": "Seed alignments for the protein families in "
```

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @Sann5, thanks for those changes! The error handling is much better now 🥇 See some more comments below.

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:159]: sorry, i was debugging and forgot to erase all of these print statements 😅

File: q2_moshpit/eggnog/_utils.py
Comment (Sann5) [reply q2_moshpit/eggnog/_utils.py:153]: @misialq in this suggestion `match` is referenced before assignment.

Comment (misialq) [reply q2_moshpit/eggnog/_utils.py:153]: Oh, of course, I see now - never mind 🤦

File: q2_moshpit/eggnog/tests/test_dbs.py
Comment (misialq) [inline q2_moshpit/eggnog/tests/test_dbs.py:50]: You don't need to assign the outputs if you don't care about them.
```suggestion
        fetch_eggnog_hmmer_db(taxon_id)
```

File: q2_moshpit/eggnog/tests/test_utils.py
Comment (misialq) [inline q2_moshpit/eggnog/tests/test_utils.py:98]: This test should probably also has some assertion(s) about the _contents_ of the created outputs. Otherwise, you really are only asserting that the mocked functions were called properly and that _some_ ProteinsDirFmt object was returned...

File: q2_moshpit/eggnog/tests/test_utils.py
Comment (misialq) [inline q2_moshpit/eggnog/tests/test_utils.py:66]: I made a comment right below about adding more content assertions and this applies here too - maybe we can at least `.validate()`?

**Review by misialq — APPROVED** [review]

**Pull request merged**
